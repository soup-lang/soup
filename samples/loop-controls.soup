x: [0, 1, 2, 3, 4, 5]

console.log('----- break -----')

x |:
    console.log($)
    break

x |:
    if $ % 5 = 4
        break
    i: $
    x |:
        if $ % 4 = 3
            break
        console.log('SS', i, $)

x |:
    if $ % 5 = 3
        break
    setTimeout(%, 1)
    i: $
    x |:
        if $ % 4 = 3
            break
        console.log('AS', i, $)

x |:
    if $ % 5 = 4
        break
    i: $
    x |:
        if $ % 4 = 2
            break
        setTimeout(%, 1)
        console.log('SA', i, $)

x |:
    setTimeout(%, 1)
    if $ % 5 = 3
        break
    i: $
    x |:
        if $ % 4 = 2
            break
        console.log('AA', i, $)
        setTimeout(%, 1)

console.log('---- in func ----')

func aab(%%)
    x |:
        i: $
        setTimeout(%, 1)
        x |:
            console.log('FAA', i, $)
            setTimeout(%, 1)
            if $ % 4 = 2
                break
        if $ % 5 = 2
            break

func sab(%%)
    x |:
        i: $
        x |:
            console.log('FSA', i, $)
            setTimeout(%, 1)
            if $ % 4 = 3
                break
        if $ % 5 = 1
            break

aab(%%)
sab(%%)

console.log('---- continue ---')

x |:
    if $ % 3 = 2
        continue
    i: $
    x |:
        if $ % 2 = 1
            continue
        console.log('SS', i, $)

x |:
    if $ % 3 = 0
        continue
    i: $
    x |:
        setTimeout(%, 1)
        if $ % 2 = 0
            continue
        console.log('SA', i, $)

x |:
    if $ % 3 = 2
        continue
    setTimeout(%, 1)
    i: $
    x |:
        if $ % 2 = 1
            continue
        console.log('AS', i, $)

x |:
    setTimeout(%, 1)
    if $ % 3 = 1
        continue
    i: $
    x |:
        if $ % 2 = 0
            continue
        console.log('AA', i, $)
        setTimeout(%, 1)

console.log('------ end ------')
