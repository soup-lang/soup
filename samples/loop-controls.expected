----- break -----
0
SS 0 0
SS 0 1
SS 0 2
SS 1 0
SS 1 1
SS 1 2
SS 2 0
SS 2 1
SS 2 2
SS 3 0
SS 3 1
SS 3 2
AS 0 0
AS 0 1
AS 0 2
AS 1 0
AS 1 1
AS 1 2
AS 2 0
AS 2 1
AS 2 2
SA 0 0
SA 0 1
SA 1 0
SA 1 1
SA 2 0
SA 2 1
SA 3 0
SA 3 1
AA 0 0
AA 0 1
AA 1 0
AA 1 1
AA 2 0
AA 2 1
---- in func ----
FAA 0 0
FAA 0 1
FAA 0 2
FAA 1 0
FAA 1 1
FAA 1 2
FAA 2 0
FAA 2 1
FAA 2 2
FSA 0 0
FSA 0 1
FSA 0 2
FSA 0 3
FSA 1 0
FSA 1 1
FSA 1 2
FSA 1 3
---- continue ---
SS 0 0
SS 0 2
SS 0 4
SS 1 0
SS 1 2
SS 1 4
SS 3 0
SS 3 2
SS 3 4
SS 4 0
SS 4 2
SS 4 4
SA 1 1
SA 1 3
SA 1 5
SA 2 1
SA 2 3
SA 2 5
SA 4 1
SA 4 3
SA 4 5
SA 5 1
SA 5 3
SA 5 5
AS 0 0
AS 0 2
AS 0 4
AS 1 0
AS 1 2
AS 1 4
AS 3 0
AS 3 2
AS 3 4
AS 4 0
AS 4 2
AS 4 4
AA 0 1
AA 0 3
AA 0 5
AA 2 1
AA 2 3
AA 2 5
AA 3 1
AA 3 3
AA 3 5
AA 5 1
AA 5 3
AA 5 5
------ end ------
